minitask/new_import_data_v2.py

```diff
@@ -0,0 +1,162 @@
+
+import pandas as pd
+import requests
+import json
+import gzip
+from lxml import etree
+
+import wikipedia, gzip, threading
+from urllib.request import urlopen
+from elasticsearch import Elasticsearch as es
+from wikipedia.exceptions import PageError, DisambiguationError
+import ssl
+from multiprocessing import Pool
+import multiprocessing
+
+from elasticsearch import Elasticsearch as es
+
+from elasticsearch import helpers as h
+
+from dataclasses import dataclass
+
+@dataclass
+class Abstract:
+    """Wikipedia abstract"""
+    ID: int
+    title: str
+    abstract: str
+    url: str
+
+    @property
+    def fulltext(self):
+        return ' '.join([self.title, self.abstract])
+
+"""
+Need to create the index first before using this script
+curl -XPUT https://localhost:9200/example3 --insecure -u admin:admin
+"""
+URL = 'https://localhost:9200/example3/_doc/'
+
+# the code with idea inspired by https://www.cnblogs.com/shaosks/p/7592229.html
+
+def start_import(test_size = 1000):
+    # u6250082 Xuguang Song
+    '''import all the data news from data set and use 1000 as test set'''
+
+    data_set = []
+    with gzip.open('../../wikidump/enwiki-20210820-abstract.xml.gz', 'rb') as f:
+        doc_id = 1
+        # iterparse will yield the entire `doc` element once it finds the
+        # closing `</doc>` tag
+        for _, element in etree.iterparse(f, events=('end',), tag='doc'):
```

Comment:
**[Required, Performance]** There is no reason why this can't be done asynchronously. This can be fed to an `imap_unordered` with a thread pool. Highly advise a refactor to something like the following:

```python
def process(xmldoc):
    # ... etc
    data = {"link": xmldoc.findtext('./title'), "title": xmldoc.findtext('./url'), "art": xmldoc.findtext('./abstract')}
    index_elastic_search(data)
    xmldoc.clear()

# ... etc more code ...

with gzip.open('../../wikidump/enwiki-20210820-abstract.xml.gz', 'rb') as f:
    total_count = 0
    xml_iterator = etree.iterparse(f, events=('end',), tag='doc')
    pool = Pool(10)
    result_iter = pool.imap_unordered(process, xml_iterator , chunksize=50)
    for _ in result_iter:
        total_count += 1
    # ... etc
```